### PR TITLE
Smile 1276 extract scopus main title

### PR DESCRIPTION
**Pull request by anetteOlli**

### CODE REVIEW

File: create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java [hunk 1]
```diff
@@ -86,6 +95,45 @@ private Reference generateReference(DocTp docTp) {
...
+    }
+
+    private String getContentWithXmlTags(List<Object> contents) {
+        StringBuilder contentStringBuilder = new StringBuilder();
+        contents.forEach(content -> {
+            if (content instanceof String) {
+                contentStringBuilder.append(content);
+            } else if (content instanceof JAXBElement<?>) {
+                var element = (JAXBElement<?>) content;
+                String contentText = getContent(element).stream().map(Object::toString).collect(Collectors.joining());
```
Comment (anetteOlli) [inline create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java:119]: Finnes det en bedre måte å gjøer dette på? Synes Object::toString() høres så unødvendig ut, men fant ingen annen metode.

File: create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java [hunk 2]
```diff
@@ -86,6 +95,45 @@ private Reference generateReference(DocTp docTp) {
...
+        return docTp.getItem().getItem().getBibrecord().getHead().getCitationTitle().getTitletext();
+    }
+
+    private String extractMainTitle(DocTp docTp) {
+        StringBuilder mainTitleBuilder = new StringBuilder();
+        getTitleText(docTp)
+            .stream()
+            .filter(titleTextTp -> titleTextTp.getOriginal().equals(YesnoAtt.Y))
+            .findFirst()
+            .ifPresent(part -> mainTitleBuilder.append(getContentWithXmlTags(part.getContent())));
```
Comment (gregorgabriel) [inline create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java:108]: Hvorfor vil vi har "content with xml tags"? Er det ikke en virkelig stygt tittel string etterhvert?

Comment (anetteOlli) [reply create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java:108]: Jo, det er en virkelig stygg tittel, vedlagte test har en ganske stygg tittel.
Vært en lang diskusjon på hvordan dette skal gjøres, med både frontendere, backendere og etterhvert PO.

Konklusjonen var å gjøre det inkrementelt: Først ta vare på xml-tagsene, slik at vi ikke mister informasjon.
Neste omgang hvis frontend får til å parse LatEx som er målet, så skal xml tagsene string konverteres til LaTex (stor task).

File: create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java [hunk 3]
```diff
@@ -86,6 +95,45 @@ private Reference generateReference(DocTp docTp) {
...
+                contentStringBuilder.append(content);
+            } else if (content instanceof JAXBElement<?>) {
+                var element = (JAXBElement<?>) content;
+                String contentText = getContent(element).stream().map(Object::toString).collect(Collectors.joining());
+                contentStringBuilder.append(String.format(XML_TAG_BUILDER, element.getName(),
+                                                          contentText, element.getName()));
+            }
+        });
+        return contentStringBuilder.toString();
+    }
```
Comment (brinxmat) [inline create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java:125]: We should have a short conversation about this:
 - Not sure I have understood what is happening here
 - Avoid lambdas with method bodies (for readability)
 - Where we have if - else if, in this and the called method, are we ignoring certain cases?

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java [hunk 4]
```diff
@@ -86,6 +95,45 @@ private Reference generateReference(DocTp docTp) {
...
 
+    private List<TitletextTp> getTitleText(DocTp docTp) {
+        return docTp.getItem().getItem().getBibrecord().getHead().getCitationTitle().getTitletext();
+    }
+
+    private String extractMainTitle(DocTp docTp) {
+        StringBuilder mainTitleBuilder = new StringBuilder();
+        getTitleText(docTp)
+            .stream()
+            .filter(titleTextTp -> titleTextTp.getOriginal().equals(YesnoAtt.Y))
```
[the remark this answers is not in the document]
Comment (anetteOlli) [reply create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java]: Swapped to marshal object to xml.

File: create-publication-from-scopus/src/test/java/no/sikt/nva/scopus/ScopusHandlerTest.java [hunk 5]
```diff
@@ -52,6 +53,13 @@
     private static final String SCOPUS_XML_0000469852 = "2-s2.0-0000469852.xml";
     private static final String SCP_ID_IN_0000469852 = "0000469852";
     private static final String DOI_IN_0000469852 = "10.1017/S0960428600000743";
+    private static final String SCOPUS_XML_0000833530 = "2-s2.0-0000833530.xml";
+    private static final String XML_ENCODING_DECLARATION = "<?xml version=\"1.0\" encoding=\"UTF-8\" "
+                                                           + "standalone=\"yes\"?>";
+    private static final String HARDCODED_EXPECTED_TITLE_IN_0000833530 = "Measurement of A\n"
+                                                                         + "    <sup>bb</sup>\n"
+                                                                         + "    <inf>FB</inf> in hadronic Z decays "
+                                                                         + "using a jet charge technique</titletextTp>";
```
Comment (anetteOlli) [inline create-publication-from-scopus/src/test/java/no/sikt/nva/scopus/ScopusHandlerTest.java:62]: There alot more information inside the xml output. Are we intrested in knowing what version of the xjd files are at compile time? I don't thinks so, so I didn't include it.

File: create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java [hunk 6]
```diff
@@ -86,6 +95,45 @@ private Reference generateReference(DocTp docTp) {
...
+            .findFirst()
+            .ifPresent(part -> mainTitleBuilder.append(getContentWithXmlTags(part.getContent())));
+        return mainTitleBuilder.toString();
```
Comment (axthosarouris) [inline create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java]: It is a bit hidden what happens when there is no original title. You could avoid it like this (but better to use the SingletonCollector)
```
    private static final String EMPTY_STIRING = "";

private String extractMainTitle(DocTp docTp) {
        return getTitleText(docTp)
            .stream()
            .filter(this::isOriginalTitle)
            .findAny()
            .map(RichstringWithMMLType::getContent)
            .map(this::getContentWithXmlTags)
            .orElse(EMPTY_STIRING);
    }
```

File: create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java [hunk 7]
```diff
@@ -86,6 +95,45 @@ private Reference generateReference(DocTp docTp) {
...
+    }
+
+    private String getContentWithXmlTags(List<Object> contents) {
+        StringBuilder contentStringBuilder = new StringBuilder();
+        contents.forEach(content -> {
+            if (content instanceof String) {
+                contentStringBuilder.append(content);
+            } else if (content instanceof JAXBElement<?>) {
+                var element = (JAXBElement<?>) content;
```
Comment (axthosarouris) [inline create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java]: extract to method with derscriptive name. e.g. converXmlElementFoundInFreeTextToString

File: create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java (hunk 6 above)
Comment (anetteOlli) [reply create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java]: There should be fresher changes than this. (amongst other this should have been fixed)

File: create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java (hunk 7 above)
Comment (anetteOlli) [reply create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java]: There should be fresher changes than this. (amongst other this should have been fixed)

File: create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java [hunk 8]
```diff
@@ -86,6 +95,45 @@ private Reference generateReference(DocTp docTp) {
...
+            .stream()
+            .filter(titleTextTp -> titleTextTp.getOriginal().equals(YesnoAtt.Y))
+            .findFirst()
+            .ifPresent(part -> mainTitleBuilder.append(getContentWithXmlTags(part.getContent())));
+        return mainTitleBuilder.toString();
+    }
+
+    private String getContentWithXmlTags(List<Object> contents) {
+        StringBuilder contentStringBuilder = new StringBuilder();
+        contents.forEach(content -> {
```
Comment (axthosarouris) [inline create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java]: Suggestion:
```
 private String getContentWithXmlTags(List<Object> contents) {
        return contents.stream()
            .map(this::convertToStringApplicableForTextValue)
            .collect(Collectors.joining());
    }

    private String convertToStringApplicableForTextValue(Object content) {
        if (content instanceof String) {
            return (String) content;
        } else if (content instanceof JAXBElement<?>) {
            return convertXmlElementFoundInFreeTextToString((JAXBElement<?>) content);
        } else {
            throw new UnsupportedOperationException(errorMessageForUnsupportedContentInFreeTextValue(content));
        }
    }

    private String errorMessageForUnsupportedContentInFreeTextValue(Serializable content) {
        return String.format("Content of type %s is not supported yet", content.getClass());
    }

    private String convertXmlElementFoundInFreeTextToString(JAXBElement<?> content) {
        String contentText = getContent(content).stream().map(Object::toString).collect(Collectors.joining());
        return String.format(XML_TAG_BUILDER, content.getName(), contentText, content.getName());
    }
```

File: create-publication-from-scopus/src/test/java/no/sikt/nva/scopus/ScopusHandlerTest.java [hunk 9]
```diff
@@ -93,6 +98,16 @@ void shouldExtractDoiAndPlaceItInsideReferenceObject() throws IOException {
...
     }
 
+    @Test
+    void shouldExtractMainTitleAndPlaceItInsideEntityDescriptionObject() throws IOException {
+        var scopusFile = IoUtils.stringFromResources(Path.of(SCOPUS_XML_0000833530));
+        var uri = s3Driver.insertFile(UnixPath.of(randomString()), scopusFile);
+        S3Event s3Event = createS3Event(uri);
+        CreatePublicationRequest createPublicationRequest = scopusHandler.handleRequest(s3Event, CONTEXT);
+        String actualMainTitle = createPublicationRequest.getEntityDescription().getMainTitle();
+        assertEquals(HARDCODED_EXPECTED_TITLE_IN_0000833530, actualMainTitle);
```
Comment (axthosarouris) [inline create-publication-from-scopus/src/test/java/no/sikt/nva/scopus/ScopusHandlerTest.java]: This test does not test only the placement of the title. It also tests that XML elements are serialized properly. This should be another test.

**Review by axthosarouris — COMMENTED** [review]
see comments above

File: create-publication-from-scopus/src/test/java/no/sikt/nva/scopus/ScopusHandlerTest.java (hunk 9 above)
Comment (anetteOlli) [reply create-publication-from-scopus/src/test/java/no/sikt/nva/scopus/ScopusHandlerTest.java]: There should be fresher changes than this. (amongst other this should have been fixed)

File: create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java [hunk 10]
```diff
@@ -86,6 +91,33 @@ private Reference generateReference(DocTp docTp) {
...
+            .stream()
+            .filter(this::isTitleOriginal)
+            .findFirst();
+    }
+
+    private boolean isTitleOriginal(TitletextTp titletextTp) {
+        return titletextTp.getOriginal().equals(YesnoAtt.Y);
+    }
+
+    private String marshalMainTitleToXmlStringInOrderToPreserveUnderlyingSupAndInfTags(TitletextTp contents) {
```
Comment (brinxmat) [inline create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java:115]: marhalMainTitleToXmlPreservingUnderlyingStructure (we don't explicitly exclude anything, right?)

File: create-publication-from-scopus/src/test/java/no/sikt/nva/scopus/ScopusHandlerTest.java (hunk 5 above)
Comment (brinxmat) [reply create-publication-from-scopus/src/test/java/no/sikt/nva/scopus/ScopusHandlerTest.java:62]: We are probably interested in the namespaces

File: create-publication-from-scopus/src/test/java/no/sikt/nva/scopus/ScopusHandlerTest.java [hunk 11]
```diff
@@ -93,6 +101,17 @@ void shouldExtractDoiAndPlaceItInsideReferenceObject() throws IOException {
         assertThat(createPublicationRequest.getEntityDescription().getReference().getDoi(), equalToObject(expectedURI));
     }
 
+    @Test
+    void shouldExtractMainTitleAndPlaceItInsideEntityDescriptionObject() throws IOException {
```
Comment (brinxmat) [inline create-publication-from-scopus/src/test/java/no/sikt/nva/scopus/ScopusHandlerTest.java:105]: shouldReturnCreatePublicationRequestWithMainTitle is good enough for me, but opinion may vary :D

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java (hunk 10 above)
Comment (anetteOlli) [reply create-publication-from-scopus/src/main/java/no/sikt/nva/scopus/ScopusHandler.java]: Fixed

File: create-publication-from-scopus/src/test/java/no/sikt/nva/scopus/ScopusHandlerTest.java (hunk 11 above)
Comment (anetteOlli) [reply create-publication-from-scopus/src/test/java/no/sikt/nva/scopus/ScopusHandlerTest.java]: Fixed

File: create-publication-from-scopus/src/test/java/no/sikt/nva/scopus/ScopusHandlerTest.java (hunk 5 above)
Comment (anetteOlli) [reply create-publication-from-scopus/src/test/java/no/sikt/nva/scopus/ScopusHandlerTest.java:65]: Fixed

**Review by axthosarouris — APPROVED** [review]

**Review by brinxmat — APPROVED** [review]

**Pull request merged**